File: common/src/main/scala/com/taskadapter/model/GTaskBuilder.scala
package com.taskadapter.model

import java.util.{Calendar, Date}

import scala.util.Random

object GTaskBuilder {
  def withRandom(fieldName: String): GTask = {
    new GTaskBuilder().withRandom(Field(fieldName)).build()
  }

  def withRandom(field: Field[_]): GTask = {
    new GTaskBuilder().withRandom(field).build()
  }
}

class GTaskBuilder {
  val task = new GTask

  def withField[T](field: Field[T], value: T): GTaskBuilder = {
    task.setValue(field, value)
    this
  }

  def withRandom(field: Field[_]): GTaskBuilder = {
    field match {
      case x: CustomDate => task.setValue(x, getDateRoundedToMinutes)
      case x: CustomFloat => val value = Random.nextFloat() * 100
        //       round to 2 digits
        val double = Math.round(value * 100.0) / 100.0
        task.setValue(x, double.floatValue())
      //      case "GUser" => new GUser(null, Random.nextString(3), Random.nextString(10))
      //      case "String" => "value " + new Date().getTime
      case Summary => task.setValue(Summary, randomStr())
      case Description => task.setValue(Description, randomStr())
      case Assignee => task.setValue(Assignee, new GUser(null, Random.nextString(3), Random.nextString(10)))
      case x: CustomString => task.setValue(x, "value " + new Date().getTime)
    }
    this
  }

  def build(): GTask = {
    task
  }

  def getDateRoundedToMinutes: Date = {
    val cal = Calendar.getInstance
    cal.set(Calendar.SECOND, 0)
    cal.set(Calendar.MILLISECOND, 0)
    cal.getTime
  }

  def randomStr(): String = "value " + new Date().getTime
}

File: test-utilities/src/main/java/com/taskadapter/model/GTaskBuilder.java
package com.taskadapter.model;

import com.taskadapter.connector.testlib.RandomStringGenerator;

import java.util.Calendar;
import java.util.Date;
import java.util.List;
import java.util.Random;

public class GTaskBuilder {
    private final GTask task = new GTask();

    public static GTask gtaskWithRandom(String fieldName) {
        return new GTaskBuilder().withRandom(new CustomString(fieldName)).build();
    }

    public static GTask gtaskWithRandomJava(Field<?> field) {
        return new GTaskBuilder().withRandom(field).build();
    }

    public static GTask gtaskWithRandom(Field<?> field) {
        return new GTaskBuilder().withRandom(field).build();
    }

    public static GTask withSummary(String value) {
        return new GTaskBuilder().withField(AllFields.summary, value).build();
    }

    public static GTask withSummary() {
        var value = new Random().nextDouble() + "";
        return withSummary(value);
    }

    public static List<GTask> getTwo() {
        return List.of(withSummary(), withSummary());
    }

    public <T> GTaskBuilder withField(Field<T> field, T value) {
        task.setValue(field, value);
        return this;
    }

    public GTaskBuilder withAssigneeLogin(String loginName) {
        task.setValue(AllFields.assigneeLoginName, loginName);
        return this;
    }

    public GTaskBuilder withRandom(Field<?> field) {
        if (field instanceof CustomDate) {
            task.setValue((CustomDate) field, getDateRoundedToMinutes());
            return this;
        }

        if (field instanceof CustomFloat) {
            var value = new Random().nextFloat() * 100;
            //       round to 2 digits
            var doubleValue = Math.round(value * 100.0) / 100.0;
            task.setValue((CustomFloat) field, (float) doubleValue);
            return this;
        }
        //      case "GUser" -> new GUser(null, Random.nextString(3), Random.nextString(10))
        //      case "String" -> "value " + new Date().getTime

        if (field.equals(AllFields.summary)) {
            task.setValue(AllFields.summary, randomStr());
            return this;
        }
        if (field.equals(AllFields.description)) {
            task.setValue(AllFields.description, randomStr());
            return this;
        }
        if (field.equals(AllFields.assigneeLoginName)) {
            task.setValue(AllFields.assigneeLoginName, RandomStringGenerator.randomAlphaNumeric(10));
            return this;
        }
        if (field.equals(AllFields.assigneeFullName)) {
            task.setValue(AllFields.assigneeFullName, RandomStringGenerator.randomAlphaNumeric(10));
            return this;
        }

        if (field instanceof CustomString) {
            task.setValue((CustomString) field, "value " + new Date().getTime());
            return this;
        }
        throw new RuntimeException("unknown field type: " + field.getClass());
    }

    public GTask build() {
        return task;
    }

    public static Date getDateRoundedToMinutes() {
        var cal = Calendar.getInstance();
        cal.set(Calendar.SECOND, 0);
        cal.set(Calendar.MILLISECOND, 0);
        return cal.getTime();
    }

    public static String randomStr() {
        return "value " + new Date().getTime();
    }
}
